engine/apps/alerts/escalation_snapshot/escalation_snapshot_mixin.py

```diff
@@ -87,8 +84,7 @@ def build_raw_escalation_snapshot(self) -> dict:
             'next_step_eta': '2021-10-18T10:28:28.890369Z
         }
         """
-
-        escalation_snapshot = None
+        data = {}
```

Comment:
this means we will no longer be setting `AlertGroup.raw_escalation_snapshot` to `None`. Instead:
```python3
>>> from apps.alerts.escalation_snapshot.snapshot_classes import EscalationSnapshot
>>> EscalationSnapshot.serializer({}).data
{'channel_filter_snapshot': None, 'escalation_chain_snapshot': None, 'last_active_escalation_policy_order': None, 'escalation_policies_snapshots': [], 'slack_channel_id': None, 'pause_escalation': False, 'next_step_eta': None}
```

Reply:
Please check if there will be no issues in places where we check `raw_escalation_snapshot` or `escalation_snapshot` since it also won't be `None`.
For example, we may encounter issues in escalate alert group and incident log builder (not sure, need to check)

Reply:
I've added a suite of tests for the `EscalationSnapshotMixin` class ([commit](https://github.com/grafana/oncall/pull/1266/commits/21424c15c7bab84d9dec99d276e4ca441d14cf1e)), so all should be good there.

But it seems there are a few spots that would need to be further refactored:
- [`apps/alerts/incident_log_builder/incident_log_builder.py`](https://github.com/grafana/oncall/blob/dev/engine/apps/alerts/incident_log_builder/incident_log_builder.py#L124-L240)
- [`apps/alerts/tasks/escalate_alert_group.py`](https://github.com/grafana/oncall/blob/dev/engine/apps/alerts/tasks/escalate_alert_group.py#L62-L82)
- [`apps/alerts/tasks/notify_all.py`](https://github.com/grafana/oncall/blob/dev/engine/apps/alerts/tasks/notify_all.py#L26-L81)
- [`apps/alerts/tasks/notify_group.py`](https://github.com/grafana/oncall/blob/dev/engine/apps/alerts/tasks/notify_group.py#L37-L74)

The main thing is how to properly refactor the checks like:
```python
if escalation_snapshot is not None:
   # do something
```

@Ferril would you mind giving some guidance on this (or hoping in to help refactor these few spots? 😄 )

Reply:
In the case `incident_log_builder.py`:
we can check if there are any escalation policies in escalation snapshot instead. By this checking we avoid unnecessary requests to db.

In `escalate_alert_group.py`:
if we remove this check, escalation should go like there are just no escalation policies. I'm not sure if we need to make any distinction between "no escalation policies" and "no escalation chain". If it is important, we can check if `escalation_chain` in the snapshot is not `None`. The difference is in additional log and `is_escalation_finished` flag.

In cases
`apps/alerts/tasks/notify_all.py`
`apps/alerts/tasks/notify_group.py`
it looks like this check doesn't make any sense now because we work with escalation snapshot like it is not `None` a few lines above in the same task 🤔

Reply:
for the `incident_log_builder.py` case, this should be addressed in [this commit](https://github.com/grafana/oncall/pull/1266/commits/992200f2f0468761e44612aec0fbcd489716843f).

If I understand your comments correctly on the other three files, those _should_ be safe?

Reply:
Right 👍